File: src/main/java/com/github/windchopper/tools/process/browser/ProcessInfo.java
package com.github.windchopper.tools.process.browser;

import com.github.windchopper.common.fx.CellFactories;
import javafx.beans.property.ReadOnlyStringWrapper;
import javafx.beans.value.ObservableValue;
import javafx.scene.control.TableCell;
import javafx.scene.control.TableColumn;
import javafx.util.Callback;

import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.Optional;

public class ProcessInfo {

    private final long pid;
    private final long parentPid;
    private final String name;
    private final String command;

    public ProcessInfo(ProcessHandle processHandle) {
        pid = processHandle.pid();
        parentPid = processHandle.parent()
            .map(ProcessHandle::pid)
            .orElse(-1L);
        name = processHandle.info().command()
            .map(Paths::get)
            .map(Path::getFileName)
            .map(Object::toString)
            .orElse("?");
        command = processHandle.info().command()
            .orElse("?");
    }

    public long pid() {
        return pid;
    }

    public long parendPid() {
        return parentPid;
    }

    public String name() {
        return name;
    }

    public String command() {
        return command;
    }

    public void destroyForcibly() {
        ProcessHandle.of(pid)
            .ifPresent(ProcessHandle::destroyForcibly);
    }

    public static Callback<TableColumn<ProcessInfo, String>, TableCell<ProcessInfo, String>> tableCellFactory() {
        return CellFactories.tableColumnCellFactory((cell, column, item, empty) -> cell.setText(item));
    }

    public static Callback<TableColumn.CellDataFeatures<ProcessInfo, String>, ObservableValue<String>> tableCellValueFactory() {
        return (features) -> switch (features.getTableColumn().getId()) {
            default -> null;
            case "identifierColumn" -> new ReadOnlyStringWrapper(Optional.ofNullable(features.getValue())
                .map(ProcessInfo::pid)
                .map(Object::toString)
                .orElse(null));
            case "parentIdentifierColumn" -> new ReadOnlyStringWrapper(Optional.ofNullable(features.getValue())
                .map(ProcessInfo::parendPid)
                .filter(pid -> pid >= 0)
                .map(Object::toString)
                .orElse(null));
            case "nameColumn" -> new ReadOnlyStringWrapper(Optional.ofNullable(features.getValue())
                .map(ProcessInfo::name)
                .orElse(null));
            case "executablePathColumn" -> new ReadOnlyStringWrapper(Optional.ofNullable(features.getValue())
                .map(ProcessInfo::command)
                .orElse(null));
        };
    }

}

File: src/main/java/com/github/windchopper/tools/process/browser/ProcessInfo.kt
package com.github.windchopper.tools.process.browser

import com.github.windchopper.common.fx.CellFactories
import javafx.beans.property.ReadOnlyStringWrapper
import javafx.beans.property.StringProperty
import javafx.beans.value.ObservableValue
import javafx.scene.control.TableCell
import javafx.scene.control.TableColumn
import javafx.scene.control.TableColumn.CellDataFeatures
import javafx.util.Callback
import java.nio.file.Paths

class ProcessInfo(processHandle: ProcessHandle) {

    val pid: Long = processHandle.pid()
    private val parentPid: Long? = processHandle.parent().map { it.pid() }.orElse(null)
    val name: String = processHandle.info().command().map { Paths.get(it) }.map { it.fileName }.map { "${it}" }.orElse("?")
    val command: String = processHandle.info().command().orElse("?")

    private val pidProperty: StringProperty = ReadOnlyStringWrapper("${pid}")
    private val parentPidProperty: StringProperty = ReadOnlyStringWrapper(parentPid?.let { "${it}" }?:"")
    private val nameProperty: StringProperty = ReadOnlyStringWrapper(name)
    private val commandProperty: StringProperty = ReadOnlyStringWrapper(command)

    fun destroyForcibly() {
        ProcessHandle.of(pid)
            .ifPresent { it.destroyForcibly() }
    }

    companion object {

        @JvmStatic fun tableCellFactory(): Callback<TableColumn<ProcessInfo?, String?>, TableCell<ProcessInfo?, String?>> {
            return CellFactories.tableColumnCellFactory { cell, column, item, empty -> cell.setText(item) }
        }

        @JvmStatic fun tableCellValueFactory(): Callback<CellDataFeatures<ProcessInfo, String>, ObservableValue<String>> {
            return Callback {
                when (it.tableColumn.id) {
                    "identifierColumn" -> it.value.pidProperty
                    "parentIdentifierColumn" -> it.value.parentPidProperty
                    "nameColumn" -> it.value.nameProperty
                    "executablePathColumn" -> it.value.commandProperty
                        else -> throw IllegalArgumentException("Unknown column: ${it.tableColumn.id}")
                }
            }
        }

    }

}